### PR TITLE
AWS HTTP handler

### PR DESCRIPTION
**Pull request by chulanovskyi-bs**
This PR introduces the `HttpHandler` module, that's specifically expected by different AWS-based JS clients such as [glue](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/glue/), [redshift](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/redshift/) and others, as an optional `requestHandler` parameter

### CODE REVIEW

File: src/awsHttpHandler/httpHandler.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import { hckFetch } from '../index';
+import { requestTimeout } from './requestTimeout';
+import { isEmpty } from './isEmpty';
+
+type LogHandler = (...content: unknown[]) => string;
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: Log functions are typically `void`.

Moreover, I think that you should clarify the type of the parameters to be able to safely invoke the function: `...content: unknown[]` is too vague. Remember that the library defines the contract that the callee needs to honour, not the opposite

```suggestion
type LogFn = (message: string, ...params?: unknown[]) => void;
```

File: src/awsHttpHandler/httpHandler.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+export type Logger = {
+  log?: LogHandler;
+  info?: LogHandler;
+};
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: I think that you should not have `.log()` and rather define one method per level. Consumers that have just a `.log()` method can implement all levels by calling that same function. 

```suggestion
export type Logger = {
  debug?: LogFn;
  info?: LogFn;
  warn?: LogFn;
};
```
Note that a library is supposed to throw errors, not to log them: that's why there is no `error` method.

File: src/awsHttpHandler/httpHandler.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+export type HttpHandlerConstructor = {
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]:
```suggestion
export type HttpHandlerConstructorParams = {
```

File: src/awsHttpHandler/httpHandler.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+type RequestOptions = {
+  headers: Headers;
+  body?: any;
+  method: string;
+  signal?: AbortSignal;
+};
+
+export class HttpHandler {
+  requestTimeout?: HttpHandlerConstructor['requestTimeout'];
+  logger?: Logger;
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: I suggest setting the logger to a no-op version by default (aka log functions that exist but do nothing). That way, you don't need to deal with the logger being `undefined` later in your logic.

```suggestion
  logger: Logger;
```

File: src/awsHttpHandler/httpHandler.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      method,
+    };
+
+    if (typeof AbortController !== 'undefined') {
+      requestOptions.signal = abortSignal;
+    }
+
+    const fetchRequest = new Request(url, requestOptions);
+
+    this.log('info', { fetchRequest }, 'Http request details');
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]:
```suggestion
    this.logger.info('Http request details', { fetchRequest });
```

File: src/awsHttpHandler/httpHandler.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      const response = {
+        headers: transformedHeaders,
+        statusCode: result.status,
+        body: hasReadableStream ? result.body : await result.blob(),
+      };
+
+      return { response };
+    };
+
+    const race = [hckFetch(fetchRequest).then(resultResolver), requestTimeout({ requestTimeout: this.requestTimeout })];
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: `AbortController` is widely available now: see [here](https://caniuse.com/?search=AbortController). I don't think that it is necessary to create so much complexity for graceful degradation with `Promise.race`...

File: src/awsHttpHandler/httpHandler.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    return Promise.race(race);
+  }
+
+  log(...payload: Parameters<LogHandler>) {
+    const method = this.logger?.log || this.logger?.info;
+    if (method) {
+      method(payload);
+    }
+  }
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: With the improvements that I described before, this method can be removed.

```suggestion
```

File: src/awsHttpHandler/requestTimeout.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  return new Promise((_, reject) => {
+    if (requestTimeout) {
+      setTimeout(() => {
+        const timeoutError = new Error(`Request did not complete within ${requestTimeout} ms`);
+        timeoutError.name = 'TimeoutError';
+        reject(timeoutError);
+      }, requestTimeout);
+    }
+  });
+}
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/requestTimeout.ts]: As explained, you can probably get rid of this module.

```suggestion
```

File: src/awsHttpHandler/httpHandler.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+};
+
+type RequestOptions = {
+  headers: Headers;
+  body?: any;
+  method: string;
+  signal?: AbortSignal;
+};
+
+export class HttpHandler {
```
Comment (thomas-jakemeyn) [inline src/awsHttpHandler/httpHandler.ts]: You should implement [the interface from aws-sdk](https://github.com/aws/aws-sdk-js-v3/blob/5d9bc5f32980f43a9b581b1f5101cdc7f2258173/packages/fetch-http-handler/src/fetch-http-handler.ts#L20) and reuse the appropriate types.

```suggestion
export class HckFetchHttpHandler implements HttpHandler {
```

Note that the interface belongs to [@aws-sdk/protocol-http](https://www.npmjs.com/package/@aws-sdk/protocol-http) that has been deprecated in favour of [@smithy/protocol-http](https://www.npmjs.com/package/@smithy/protocol-http).

**Review by thomas-jakemeyn — CHANGES_REQUESTED** [review]
Some additional comments:

- I would rename the folder `src/awsHttpHandler` to `src/aws-sdk` as we could add extensions for others libraries in the future.
- You need to enrich the DTD file with what you expose (it is not automatic).

**Comment by chulanovskyi-bs** [comment]
@thomas-jakemeyn All remarks were addressed, please review

File: src/hckFetchAwsSdkHttpHandler.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  async handle(request: HttpRequest, { abortSignal }: HttpHandlerOptions = {}): Promise<{ response: HttpResponse }> {
+    const requestTimeoutInMs = this.options.requestTimeout;
+    const keepAlive = this.options.keepAlive === true;
+    const credentials = this.options.credentials as RequestInit['credentials'];
+
+    // if the request was already aborted, prevent doing extra work
+    if (abortSignal?.aborted) {
+      const abortError = new Error('Request aborted');
+      abortError.name = 'AbortError';
+      return Promise.reject(abortError);
```
Comment (bigorn0) [inline src/hckFetchAwsSdkHttpHandler.ts:38]: @thomas-jakemeyn shouldn't this just be a `throw abortError` ?

Comment (thomas-jakemeyn) [reply src/hckFetchAwsSdkHttpHandler.ts:38]: I sticked as much as possible to the default implementation from aws-sdk. I did not try to improve it in order to ease future evolutions.

**Review by thomas-jakemeyn — APPROVED** [review]

**Pull request merged**